### PR TITLE
HW6 Task02

### PR DESCRIPTION
**Pull request by LinaKrevetochka**

### CODE REVIEW

File: hw/hw_06_task_02.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        num of days to do homework
+    created: datetime.timedelta
+        date and time of deadline
+
+    """
+
+    def __init__(self, text: str, deadline: int):
+        """Create an instance of a homework."""
+        self.text = text
+        self.deadline = timedelta(deadline)
```
Comment (Cjkjvfnby) [inline hw/hw_06_task_02.py]: https://teams.microsoft.com/l/message/19:8cafe0713cf0416a9abeeb8b6222b712@thread.tacv2/1606083077755?tenantId=b41b72d0-4e9f-4c26-8a69-f949f367c91d&groupId=91c90c81-2cfe-4bd5-bb81-4d05e8481e0a&parentMessageId=1606083077755&teamName=EPAM%20Python%20Oct%2720&channelName=%D0%9E%D0%B1%D1%89%D0%B8%D0%B9&createdTime=1606083077755

File: hw/hw_06_task_02.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        self.created = datetime.now()
+
+    def is_active(self) -> bool:
+        """Return False if deadline is expired, True otherwise."""
+        hw_deadline = self.created + self.deadline
+        now = datetime.now()
+        return not now > hw_deadline
```
Comment (Cjkjvfnby) [inline hw/hw_06_task_02.py]: You can simplify code by reversing conditions

```
not now > hw_deadline

now <= hw_deadline

```

File: hw/hw_06_task_02.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    A class to represent a student.
+
+    Methods
+    -------
+    do_homework():
+        Return homework if deadline is available, raise exception otherwise.
+
+    """
+
+    # flake8: noqa: T001
```
Comment (Cjkjvfnby) [inline hw/hw_06_task_02.py:112]: Looks like you copy-pasted this and forget to change it.

File: hw/hw_06_task_02.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    def do_homework(self, hw: Homework, solution: str) -> Optional[Homework]:
+        """Return None and print message if deadline is passed, an instance of homework otherwise."""
+        if not hw.is_active():
```
Comment (Cjkjvfnby) [inline hw/hw_06_task_02.py]: This line has 3 actions:  condition, negation, and function call.  You can reorganize the code and get rid of the negation.

```
if hw.is_active():
    return HomeworkResult(self, hw, solution)
raise DeadlineError()

File: hw/hw_06_task_02.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+    homework_done = defaultdict(list)
+
+    def create_homework(self, text: str, deadline: int) -> Homework:
+        """Return an instance of created homework."""
+        return Homework(text, deadline)
+
+    def check_homework(self, hw: HomeworkResult) -> bool:
+        if len(hw.solution) > 5:
+            if not any(hw in s for s in self.homework_done.values()):
```
Comment (Cjkjvfnby) [inline hw/hw_06_task_02.py]: It should be unique for homework, no need to check across all homeworks.

File: test/test_02_hw_06.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+def test_do_current_homework(good_student, docs_hw):
+    hw = good_student.do_homework(docs_hw, "I have done this hw.")
+    assert isinstance(hw, HomeworkResult)
+    assert hw.author is good_student
+    assert hw.solution == "I have done this hw."
+    assert hw.homework is docs_hw
+
+
+def test_do_expired_homework(lazy_student, expired_homework):
+    with pytest.raises(DeadlineError, match="You are late."):
```
Comment (Cjkjvfnby) [inline test/test_02_hw_06.py]: Do you need to check `match`  for that exception?

File: test/test_02_hw_06.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    temp_2 = Teacher.homework_done
+    assert temp_1 == temp_2
+
+
+def test_not_a_homework_object():
+    with pytest.raises(TypeError, match="You gave a not Homework object."):
+        HomeworkResult(good_student, "fff", "Solution")
+
+
+def test_delete_hw_from_homework_done(opp_teacher, good_student, oop_hw, docs_hw):
```
Comment (Cjkjvfnby) [inline test/test_02_hw_06.py]: Do you need a teacher here?

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]

File: test/test_02_hw_06.py (hunk 7 above)
Comment (LinaKrevetochka) [reply test/test_02_hw_06.py]: Yes, I need him to check homeworks and add them to homework_done.

**Review by Cjkjvfnby — APPROVED** [review]

**Pull request merged**